plugins/lua/dwarfmonitor.lua

```diff
@@ -146,6 +146,95 @@ function Widget_cursor:render_body(p)
     p:string(self.output)
 end
 
+-- FPS counter that doesn't count frames when paused,
+-- so FPS doesn't have to slowly stabilize after unpausing.
+Widget_fps = defclass(Widget_fps, Widget)
+Widget_fps.ATTRS = {
+    output = '',
+    show_fps_in_init = df.global.gps.display_frames == 1,
+    prev_paused = false,
+    prev_clock = nil,
+    prev_frames = nil,
+    elapsed_clock = 0.0,
+    elapsed_frames = 0,
+    history = {},
+}
+function Widget_fps:fps()
+    local tps = self.elapsed_frames / self.elapsed_clock
+    local count = 1
+
+    if not (tps >= 0.0) then -- ignore NaNs
+        tps = 0
+        count = 0
+    end
+
+    for i = 1, self.opts.average_over_seconds do
+        if self.history[i] and self.history[i] >= 0.0 then -- ignore NaNs
+            tps = tps + self.history[i]
+            count = count + 1
+        end
+    end
+    return tps / count
+end
+function Widget_fps:updateOutput()
+    local fps = self:fps()
+    if not (fps >= 0.0) then -- NaN
+        self.output = "   -"
+    elseif fps <= 1.0 then -- less than 1 fps
+        self.output = string.format("%0.2f", fps)
+    else
+        self.output = string.format("%4d", math.floor(0.5 + fps))
+    end
+    self.output = "FPS:" .. self.output
+end
+function Widget_fps:update()
+    -- dont show fps counter unless [FPS:YES] is set in init.txt
+    if not self.show_fps_in_init then
+        return
+    end
+
+    -- hide internal FPS counter because it is drawn on top of this widget
+    df.global.gps.display_frames = 0
+
+    if not self.opts.average_over_seconds then
+        self.opts.average_over_seconds = 5
+    end
+
+    if df.global.pause_state then
+        self.prev_paused = true
+    else
+        local clock = os.clock()
```

Comment:
This is not portable - it returns the elapsed CPU time on Linux and macOS, not the absolute time, so I'm seeing stuff like "FPS: 342" and "FPS: 887" with a limit of 100.

Reply:
The counter is updated every second, and the value is the average of the latest five 1-second intervals of FPS. Without averaging, the FPS counter would jump unpredictable every time it was reset. I will fix it so the counter is always visible (but shows as "FPS:   -" the first second).

The counter seemed to work on my linux machine so I did not think to thoroughly understand os.clock(). 
To fix it I intend to add a variant of MiscUtils.cpp:getTimeMs64() to LuaApi.cpp:dfhack_module because lua does not provide millisecond precision timing by default. That way I can call dfhack.getTime() from lua that returns seconds since epoch with millisecond floating-point precision (similar to os.clock()).
Since lua stores numbers as 64 bit doubles, a 64 bit unsigned int will lose some precision in the year 200 million. I can live with this limitation.

And thank you very much for all your informative feedback as I get know the code base.

Reply:
Oh, was it working on Linux? I was guessing you were testing on Windows because os.clock returns absolute time there, but I guess it would effectively return absolute time on Linux if you're hitting the FPS cap.

Your idea for wrapping getTimeMs64 is good. Lua 5.3 supports 64-bit integers, and we had to switch to that for 0.43.05, so it won't lose precision until it wraps around.